JSON-Grammar-spec.md

```diff
@@ -0,0 +1,794 @@
+
+## Rationale and Portability
+
+### Why JSON-based Grammar?
+
+1. **Language Agnostic**: JSON is universally supported across programming languages, making grammars portable between different implementations.
+
+2. **Serialization**: Grammar definitions can be easily stored, transmitted, and version-controlled as standard JSON files.
+
+3. **Tooling Support**: Leverage existing JSON validation, editing, and processing tools for grammar development.
+
+4. **Interoperability**: Enable grammar sharing between different parser generators and language ecosystems.
+
+5. **Runtime Configuration**: Dynamically load and modify grammar definitions without code changes.
+
+### Portability Benefits
+
+- **Cross-platform**: JSON grammars work identically across Node.js, browsers, and other JavaScript environments
+- **Version Control**: Text-based format integrates seamlessly with Git and other VCS
+- **Documentation**: Self-documenting structure with optional metadata fields
+- **Validation**: JSON schema validation for grammar correctness
+- **Transformation**: Programmatic grammar generation and modification
+
+## Grammar File Structure
+
+A JSON grammar file defines a complete parsing grammar with the following top-level structure:
+
+```typescript
+interface Grammar {
+  start: string;                              // Entry point rule name
+  cst: Record<string, GrammarNode>;          // Concrete syntax rules
+  ast?: Record<string, AstNodeExpression>;   // AST transformation rules
+}
+```
+
+### Basic Example
+
+```json
+{
+  "start": "Value",
+  "cst": {
+    "Value": {"r": "Number"},
+    "Number": {"t": "/\\d+/"}
+  },
+  "ast": {
+    "Number": ["num", ["$", "/raw"]]
+  }
+}
+```
+
+### Structure Components
+
+#### Start Symbol
+- **`start`**: String specifying the root grammar rule name
+- Must reference a rule defined in the `cst` section
+- Determines the entry point for parsing
+
+#### CST Rules
+- **`cst`**: Object mapping rule names to grammar node definitions
+- Contains all named grammar rules that can be referenced
+- Supports all five grammar node types (Reference, Terminal, Production, Union, List)
+
+#### AST Transformations
+- **`ast`**: Optional object defining custom AST generation rules
+- Maps rule names to JSON expressions for AST transformation
+- Overrides default AST generation behavior
+
+## Grammar Node Types
+
+The JSON grammar specification supports five fundamental node types for defining parsing rules. Each type has a specific JSON representation with both full interface and shorthand syntax options.
+
+### 1. RefNode (Reference Node)
+
+References a named rule defined elsewhere in the grammar.
+
+#### Interface
+```typescript
+type RefNode<Name extends string = string> = {
+  r: Name;
+};
+```
+
+#### Syntax
+```json
+{"r": "RuleName"}
+```
+
+#### Examples
+```json
+{
+  "start": "Program",
+  "cst": {
+    "Program": {"r": "Statement"},
+    "Statement": "return;"
+  }
+}
+```
```

Comment:
Lets show example next to the grammar:


```suggestion
```js
{
  "start": "Program",
  "cst": {
    "Program": {"r": "Statement"},
    "Statement": "return;"
  }
}

// Matches:
// return;
```
```